napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -0,0 +1,257 @@
+    layer._moving_value = (None, None)
+    layer._set_highlight()
+    layer._update_thumbnail()
+
+
+def add_line_rectangle_ellipse(layer, event):
+    """Add a line, rectangle or ellipse."""
+    coord = [layer.coordinates[i] for i in layer.dims.displayed]
+
+    # on press
+    # Start drawing a rectangle / ellipse / line
+    size = layer._vertex_size * layer.scale_factor / 4
+    corner = np.array(coord)
+    if layer._mode == Mode.ADD_RECTANGLE:
+        data = np.array(
+            [corner, corner + [size, 0], corner + size, corner + [0, size]]
+        )
+        shape_type = 'rectangle'
+    elif layer._mode == Mode.ADD_ELLIPSE:
+        data = np.array(
+            [corner, corner + [size, 0], corner + size, corner + [0, size]]
+        )
+        shape_type = 'ellipse'
+    elif layer._mode == Mode.ADD_LINE:
+        data = np.array([corner, corner + size])
+        shape_type = 'line'
+    data_full = layer.expand_shape(data)
+    layer.add(data_full, shape_type=shape_type)
+    layer.selected_data = [layer.nshapes - 1]
+    layer._value = (layer.selected_data[0], 4)
+    layer._moving_value = copy(layer._value)
+    layer.refresh()
+    yield
+
+    # on move
+    while event.type == 'mouse_move':
+        # Drag any selected shapes
+        coord = [layer.coordinates[i] for i in layer.dims.displayed]
+        layer._move(coord)
+        yield
+
+    # on release
+    layer._finish_drawing()
+
+
+def add_path_polygon(layer, event):
+    """Add a path or polygon."""
+    coord = [layer.coordinates[i] for i in layer.dims.displayed]
+
+    # on press
+    if layer._is_creating is False:
+        # Start drawing a path
+        data = np.array([coord, coord])
+        data_full = layer.expand_shape(data)
+        layer.add(data_full, shape_type='path')
+        layer.selected_data = [layer.nshapes - 1]
+        layer._value = (layer.selected_data[0], 1)
+        layer._moving_value = copy(layer._value)
+        layer._is_creating = True
+        layer._set_highlight()
+    else:
+        # Add to an existing path or polygon
+        index = layer._moving_value[0]
+        if layer._mode == Mode.ADD_POLYGON:
+            new_type = Polygon
+        else:
+            new_type = None
+        vertices = layer._data_view.displayed_vertices[
+            layer._data_view.displayed_index == index
+        ]
+        vertices = np.concatenate((vertices, [coord]), axis=0)
+        # Change the selected vertex
+        layer._value = (layer._value[0], layer._value[1] + 1)
+        layer._moving_value = copy(layer._value)
+        data_full = layer.expand_shape(vertices)
+        layer._data_view.edit(index, data_full, new_type=new_type)
+        layer._selected_box = layer.interaction_box(layer.selected_data)
+
+
+def add_path_polygon_creating(layer, event):
+    """While a path or polygon move next vertex to be added."""
+    if layer._is_creating:
+        coord = [layer.coordinates[i] for i in layer.dims.displayed]
```

Comment:
this line is repeated a lot. it should be made into a helper function

Reply:
should I add a property to the layer `layer.displayed_coordinates`?

Reply:
I can put the property on the base layer and use it where needed

Reply:
change made, thanks for the suggestion - it was present in a lot of places!!!